### ISSUE TITLE
Format Feature Request: Trailing Comma for multi-line code

### ISSUE THREAD

**Issue by still-dreaming-1** [DEVSENSE/phptools-docs#384]
PHP supports trailing commas in many different parts of the language now.

There could be different options for it to support this for function parameters, function arguments, arrays, etc. I only want trailing commas in multi-line code, such as when calling a function and placing each argument on a separate line.

One advantage of standardizing on this code style is, when you add an extra argument to a multi-line function call for example, this allows the git commit and diffs to only have/show 1 line changed instead of 2.

**Comment by jakubmisek** [comment]
Thank you for the suggestion. Is it related to the code formatting feature?

**Comment by still-dreaming-1** [comment]
> Thank you for the suggestion. Is it related to the code formatting feature?

Yes. Currently when it formats a function call into multiple lines, there is no trailing comma. Also, if it formats multi-line code that is missing a trailing comma, it doesn't add one. It would be nice to have a formatter option for this.

**Comment by jakubmisek** [comment]
I think, it respects the trailing that was there previously.

Is it a PSR-something standard? :) I'm adding @Miloslav for this one.

**Comment by still-dreaming-1** [comment]
I think maybe you got a little tripped up by not focusing on the title before reading the description, I should have started from scratch in the description instead of carrying forward from the context of the title. This is just a feature request to add more options for the formatter. I believe you are correct and that it will leave a trailing comma that was previously there, I haven't encountered any bugs like it removing the comma. It's not part of PSR-12 or any other PSRs as far as I know.

Here is a [page](https://php.watch/versions/8.0/trailing-comma-parameter-use-list) that shows some of the places trailing commas are allowed. It also illustrates the cleaner diff outputs that can result from using them.

I think the key will be to have 2 separate settings for each php language feature that allows trailing commas, such that a user could have this in their settings.json:
```
php.format.rules.multiLineFunctionCallTrailingComma = true,
php.format.rules.singleLineFunctionCallTrailingComma = false,
```

When one of the settings is set to true, it should enforce trailing commas for that case, adding them if they are missing. If it is set to false, it would remove trailing commas for that case if there are any.

**Comment by still-dreaming-1** [comment]
I just wanted to point out an example of how this can be very beneficial. I use [vscode-neovim](https://github.com/vscode-neovim/vscode-neovim), and I have a mapping setup that makes it really easy and convenient for me choose to run the formatter against certain lines of code while editing. I often use this as I'm turning a single line into multiple lines. It does help and saves some work, it does everything I want except it does not add a trailing comma. So then I have to go back to the previous line and add one myself like an animal ;)

**Comment by still-dreaming-1** [comment]
Some people stand in line the night before for the new iPhone. I add new linting rules to my settings before support for them is released.

**Comment by Miloslav** [comment]
That's some serious dedication to clean code! The pre-release is out, so when you have a chance, give it a try and let us know.

**Comment by still-dreaming-1** [comment]
Nice! It seems to work well for adding commas to the end of multi-line calls. I also ended up testing it the other way around, sort of by accident. I just so happened to notice a call could be shortened to a single line. Since I'm using Vim bindings, I just pressed `J` a few times. This left it with an extra trailing comma at the end. I then tried formatting that line, and it added an extra space after the trailing comma. I was hoping it would remove the trailing comma. From my perspective, the more it can format everything the way we want the better, and I think this illustrates a real world scenario where it would be easier to have for the formatter format the trailing comma away than to do it by hand.
